src/axom/primal/geometry/NURBSCurve.hpp

```diff
@@ -0,0 +1,1177 @@
+   *
+   * \param [in] npts The target number of control points
+   * \param [in] degree The target degree
+   * 
+   * \note Will clear any data already in these arrays.
+   */
+  void setParameters(int npts, int degree)
+  {
+    SLIC_ASSERT(npts >= degree + 1);
+    SLIC_ASSERT(degree >= 0);
+
+    m_controlPoints.resize(npts);
+    makeNonrational();
+
+    m_knotvec.makeUniform(npts, degree);
+  }
+
+  /*!
+   * \brief Reset the knot vector and increase the number of control points
+   *
+   * \param [in] degree The target degree
+   * 
+   * \warning This method does NOT change the existing control points,
+   *  but will allocate the minimum (sensible) number of points needed
+   * \pre Requires degree < npts
+   */
+  void setDegree(int degree)
+  {
+    SLIC_ASSERT(degree >= 0);
+
+    int npts = getNumControlPoints();
+    SLIC_ASSERT(degree < npts);
+
+    if(npts < degree + 1)
+    {
+      npts = degree + 1;
+      m_controlPoints.resize(degree + 1);
+      m_weights.resize(degree + 1);
+    }
+
+    m_knotvec.makeUniform(npts, degree);
+  }
+
+  /// \brief Returns the degree of the NURBS Curve
+  int getDegree() const { return static_cast<int>(m_knotvec.getDegree()); }
+
+  /// \brief Returns the order of the NURBS Curve
+  int getOrder() const { return static_cast<int>(m_knotvec.getDegree() + 1); }
+
+  /// \brief Returns the number of control poitns in the NURBS Curve
+  int getNumControlPoints() const
+  {
+    return static_cast<int>(m_controlPoints.size());
+  }
+
+  /*!
+   * \brief Set the number control points
+   *
+   * \param [in] npts The target number of control points
+   * 
+   * \warning This method does NOT maintain the curve shape,
+   *  i.e. is not performing knot insertion/removal.
+   */
+  void setNumControlPoints(int npts)
+  {
+    const int deg = m_knotvec.getDegree();
+    SLIC_ASSERT(npts > deg);
+
+    m_controlPoints.resize(npts);
+    m_knotvec.makeUniform(npts, deg);
+  }
+
+  /// \brief Returns the number of knots in the NURBS Curve
+  axom::IndexType getNumKnots() const { return m_knotvec.getNumKnots(); }
+
+  /// \brief Make nonrational. If already rational, do nothing
+  void makeNonrational() { m_weights.resize(0); }
+
+  /// \brief Make trivially rational. If already rational, do nothing
+  void makeRational()
+  {
+    if(!isRational())
+    {
+      m_weights.resize(m_controlPoints.size());
+      m_weights.fill(1.0);
+    }
+  }
+
+  /// \brief Use array size as flag for rationality
+  bool isRational() const { return !m_weights.empty(); }
+
+  /// \brief Clears the list of control points, make nonrational
+  void clear()
+  {
+    m_controlPoints.clear();
+    m_knotvec.clear();
+    makeNonrational();
+  }
+
+  /// \brief Retrieve the control point at index \a idx
+  PointType& operator[](int idx) { return m_controlPoints[idx]; }
+
+  /// \brief Retrieve the control point at index \a idx
+  const PointType& operator[](int idx) const { return m_controlPoints[idx]; }
```

Comment:
Should this be an explicitly named function to access the control points (similar to `getWeight()`)

Alternatively, would it make sense to have the following API (w/ const overloads):
* `controlPoints() -> axom::Array<PointType>&`
* `weights() -> axom::Array<T>&`
* `knots() -> axom::Array<T>&`

Then users could use the underlying array functions to do things like:
```cpp
const& pt = nurbs.controlPoints()[3];
nurbs.weights()[2] = 2.3;
....
```

Note: This is different than the functions below that return a copy of the control points, knots and weights.

Reply:
I'm open to the idea of *adding* an explicitly named function, but the brackets are used in the `BezierCurve` class too, so this is just my way of keeping consistent with the old interface. On a more personal level, I think accessing control points via `[]` is fairly intuitive, and I think it's worth prioritizing the convenience of control point access vs weight/knot access.

More to your second point though, this might be the best time (or maybe last chance?) to update the old interface altogether while they aren't heavily used. For example, we access control nodes of patches with `(p, q)`, so there's already an inconsistency there. Similarly, I think adding `controlPoints()` as a type of accessor is a great idea, but I would also want to add those to `BezierCurve` and `BezierPatch`.